### PR TITLE
Made the reservation UI with Time and Date pickers functionallity

### PR DESCRIPTION
**Pull request by SofiaZarzabal**
- Reservation UI implemented, with save button which doesn't has functionality yet
- The Date Picker and Time Picker natives functionality was made

**Folder Structure**
image

**Demostration in Pixel 3a**
PIXEL 3A GIF

**Demostration in Nexus S**
NEXUS S GIF

**Unit Test**
image

**Unit Test Coverage**
image

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -9,13 +9,16 @@
...
+        <activity android:name=".activities.MainActivity">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
 
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity android:name=".activities.ParkingSpaceReservationActivity">
+
+        </activity>
```
Comment (SartiMau) [inline app/src/main/AndroidManifest.xml]: remove this close tag and just finish the line 19 with /> please

File: app/src/main/java/com/example/parkingsystemkotlin/activities/MainActivity.kt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package com.example.parkingsystemkotlin
+package com.example.parkingsystemkotlin.activities
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/activities/MainActivity.kt]: the packages shoud be in singular

File: app/src/main/java/com/example/parkingsystemkotlin/activities/ParkingSpaceReservationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityReservationParkingSpaceBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        presenter = ParkingSpaceReservationPresenter(ParkingSpaceReservationModel(), ParkingSpaceReservationView(this, binding))
+        setListeners()
+    }
+
+    companion object {
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/activities/ParkingSpaceReservationActivity.kt]: put the companion at the end of the class please

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/contracts/ParkingSpaceReservationContract.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package com.example.parkingsystemkotlin.mvp.contracts
+
+import android.app.DatePickerDialog
+import android.app.TimePickerDialog
+import java.util.*
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/contracts/ParkingSpaceReservationContract.kt]: do not use wildcard imports pls

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package com.example.parkingsystemkotlin.mvp.model
+
+import com.example.parkingsystemkotlin.mvp.contracts.ParkingSpaceReservationContract
+import com.example.parkingsystemkotlin.utils.DateUtils
+import java.text.SimpleDateFormat
+import java.util.*
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt]: same

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+class ParkingSpaceReservationModel : ParkingSpaceReservationContract.ParkingSpaceReservationModel {
+    private var isDateStartButtonPressed = false
+    private var dateStart: Calendar = Calendar.getInstance()
+    private var timeStart: Calendar = Calendar.getInstance()
+    private var dateEnd: Calendar = Calendar.getInstance()
+    private var timeEnd: Calendar = Calendar.getInstance()
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt]: it is okay to initialize all these calendars here? if the user just put save without setting the date and times the local info will be saved with the actual date/time

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    private val view: ParkingSpaceReservationContract.ParkingSpaceReservationView
+) : ParkingSpaceReservationContract.ParkingSpaceReservationPresenter {
+
+    override fun onButtonParkingSpaceReservationPickerPressed(listener: DatePickerDialog.OnDateSetListener) {
+        model.setDateStartButtonPressed(view.getButtonPickerStart())
+        view.showDatePickerDialog(listener)
+    }
+
+    override fun onDateSetPressed(year: Int, month: Int, dayOfMonth: Int, timeListener: TimePickerDialog.OnTimeSetListener) {
+        val sDate = "$dayOfMonth${Constants.SLASH}${month+Constants.ONE}${Constants.SLASH}$year"
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt]: Beautiful

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        val date = model.convertToCalendar(sDate, Constants.FORMAT_DATE)
+        if(model.getDateStartButtonPressed()){
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt]: remember to format the code properly

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
+package com.example.parkingsystemkotlin.mvp.view
+
+import android.app.Activity
+import android.app.DatePickerDialog
+import android.app.TimePickerDialog
+import com.example.parkingsystemkotlin.R
+import com.example.parkingsystemkotlin.databinding.ActivityReservationParkingSpaceBinding
+import com.example.parkingsystemkotlin.mvp.contracts.ParkingSpaceReservationContract
+import com.example.parkingsystemkotlin.mvp.view.base.ActivityView
+import java.util.*
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt]: same

File: app/src/main/java/com/example/parkingsystemkotlin/utils/DateUtils.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package com.example.parkingsystemkotlin.utils
+
+import java.text.ParseException
+import java.text.SimpleDateFormat
+import java.util.*
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/utils/DateUtils.kt]: same

File: app/src/test/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenterTest.kt [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+import android.app.DatePickerDialog
+import android.app.TimePickerDialog
+import com.example.parkingsystemkotlin.mvp.contracts.ParkingSpaceReservationContract
+import com.example.parkingsystemkotlin.mvp.model.ParkingSpaceReservationModel
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import org.junit.Before
+import org.junit.Test
+import java.util.*
```
Comment (SartiMau) [inline app/src/test/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenterTest.kt]: same

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (SofiaZarzabal) [reply app/src/main/AndroidManifest.xml]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/activities/MainActivity.kt (hunk 2 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/activities/MainActivity.kt]: Changed!

File: app/src/main/java/com/example/parkingsystemkotlin/activities/ParkingSpaceReservationActivity.kt (hunk 3 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/activities/ParkingSpaceReservationActivity.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/contracts/ParkingSpaceReservationContract.kt (hunk 4 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/contracts/ParkingSpaceReservationContract.kt]: Changed!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt (hunk 5 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt]: Changed!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt (hunk 7 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt]: Thanks!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt (hunk 8 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenter.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/utils/DateUtils.kt (hunk 10 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/utils/DateUtils.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt (hunk 9 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt]: Done!

File: app/src/test/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenterTest.kt (hunk 11 above)
Comment (SofiaZarzabal) [reply app/src/test/java/com/example/parkingsystemkotlin/mvp/presenter/ParkingSpaceReservationPresenterTest.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt (hunk 6 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/model/ParkingSpaceReservationModel.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/contract/ParkingSpaceReservationContract.kt [hunk 12]
```diff
@@ -0,0 +1,39 @@
+package com.example.parkingsystemkotlin.mvp.contract
+
+import android.app.DatePickerDialog
+import android.app.TimePickerDialog
+import android.icu.util.Calendar
+
+
```
Comment (BQuinteros) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/contract/ParkingSpaceReservationContract.kt]: pls remove extra spaces

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            binding.textParkingSpaceReservationPickerBegin.text = it.getString(R.string.text_parking_space_reservation_picker, date, time)
+        }
+    }
+
+    override fun showDateAndTimeEnd(date: String, time: String) {
+        context?.let {
+            binding.textParkingSpaceReservationPickerEnd.text = it.getString(R.string.text_parking_space_reservation_picker, date, time)
+        }
+    }
+
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt]: remove this extra lines pls

**Review by SartiMau — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt (hunk 13 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/view/ParkingSpaceReservationView.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/mvp/contract/ParkingSpaceReservationContract.kt (hunk 12 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/mvp/contract/ParkingSpaceReservationContract.kt]: Done!

**Review by SartiMau — APPROVED** [review]

**Pull request merged**